### PR TITLE
adds support for deleting init node

### PR DESCRIPTION
**Pull request by Josh-Diamond**
extensions to support: https://github.com/rancher/rancher/pull/44155

Link to previous PR (draft) for additional context around requested changes:
https://github.com/rancher/shepherd/pull/64

### CODE REVIEW

File: extensions/nodes/nodes.go
Comment (slickwarren) [inline extensions/nodes/nodes.go:49]: why not move what's in this for loop to line 36 instead? you could also break in the if statement since you're done looking for initNode.

File: extensions/nodes/nodes.go
Comment (slickwarren) [inline extensions/nodes/nodes.go:76]: log seems unnecessary

File: extensions/nodes/nodes.go
Comment (slickwarren) [inline extensions/nodes/nodes.go:78]: general question; can't you use the list function here and if list for the deleted machine returns nil, than it is deleted? (Instead of having the for loop below)

File: extensions/nodes/nodes.go
Comment (slickwarren) [inline extensions/nodes/nodes.go:25]: does listOptions have an entry for the `.Type` that you're looking for? i.e. could you just query listOptions for secret.Type=machinePlanSecretType instead of the for loop below?

**Review by slickwarren — COMMENTED** [review]
some nits/questions

File: extensions/nodes/nodes.go
Comment (susesgartner) [reply extensions/nodes/nodes.go:78]: Adding on to several of the filtering/list option things Caleb mentioned: we should always (if possible) use the List filters with API calls to simplify our logic and slightly increase performance. You will also probably need to rework your logging statements after this change.

**Review by susesgartner — COMMENTED** [review]
Looks pretty good, just add some list filtering to replace the for loops and your good to go!

**Review by markusewalker — APPROVED** [review]
LGTM

File: extensions/clusters/clusters.go
Comment (slickwarren) [inline extensions/clusters/clusters.go:1255]: this now only checks every 2 minutes, I think we should keep it to magnitude in seconds. i.e. 1 or 2 second poll

File: extensions/machinepools/machinepools.go
Comment (slickwarren) [inline extensions/machinepools/machinepools.go:313]: can you add a comment that this will always be ordered and [0] is exactly what we're expecting / the order won't change due to the label you're looking at?

**Review by slickwarren — DISMISSED** [review]

File: extensions/steve/steve.go
Comment (igomez06) [inline extensions/steve/steve.go:20]: What's the point of having this? You're just doing a delete call?

File: extensions/steve/steve.go
Comment (igomez06) [inline extensions/steve/steve.go:52]: My concern about this wait is that it's too generic and deleting resources is different depending on what that is.

**Comment by igomez06** [comment]
Josh why did you remove you original extensions for something is generic?

File: extensions/steve/steve.go
Comment (Josh-Diamond) [reply extensions/steve/steve.go:52]: too generic? 🤔 -- could you elaborate on your concern?  When would this not be applicable for awaiting a steve resource to be deleted?  I'm not understanding what you'd like to see in place of this 🙂

Comment (igomez06) [reply extensions/steve/steve.go:52]: Deleting a secret is different than deleting a cluster. We would want to create extensions to help validate those differently. Having such a generic function causes confusion for someone  writing a test

Comment (igomez06) [reply extensions/steve/steve.go:52]: Have an extensions that waits for a node to be deleted. Why did you decide to go this route than what you had before?

File: extensions/steve/steve.go
Comment (Josh-Diamond) [reply extensions/steve/steve.go:20]: I felt this was a nice enhancement to our extensions.  Anytime we'd like to delete a steve resource, (and this should be applied to all CRUD operations IMO) we can just call this extension, instead of repeating the client call + err != nil, which would make the tests cleaner, more readable and ultimately DRYer code, as we wouldn't be bloating tests with err != nil every time we want to make a client call.  That was my thinking in creating this.

Comment (igomez06) [reply extensions/steve/steve.go:20]: you still need to do a err != nil check this function returns error. Why wouldn't that happen?

File: extensions/steve/steve.go
Comment (susesgartner) [reply extensions/steve/steve.go:52]: If this covers a good chunk of the "wait for delete" use cases though wouldn't it make sense to have a generic function and then if you need something more specific you could just create it in place? @igomez06

**Comment by Josh-Diamond** [comment]
@igomez06 I enhanced my extensions as requested by reviewers to be as generic as possible, as it allows the extension to be applicable to more use cases, which is ideal when creating extensions IMO.

File: extensions/steve/steve.go
Comment (Josh-Diamond) [reply extensions/steve/steve.go:52]: Thanks for the clarification Israel!

this function just awaits the resource to be deleted, which would be applicable to secrets too.  Validating the secret was properly deleted should be a separate extension IMO.  This is purely to just await the resource deletion, and I tried to name the function appropriately, as to not cause any confusion.

I went this route as requested by reviewers.  The concern was using v3 vs v1 and the ask was to refactor out all the v3 calls.  This came about as the result of that re-work.  

If I edit this code to be less generic, we will have to add more extensions, extremely similar to each other, for each specific use case, which just seems unnecessary when it could be tackled through one extension.

Comment (igomez06) [reply extensions/steve/steve.go:52]: You make a good point though Sam, I'll drop it.

**Comment by igomez06** [comment]
> @igomez06 I enhanced my extensions as requested by reviewers to be as generic as possible, as it allows the extension to be applicable to more use cases, which is ideal when creating extensions IMO.

Sure Josh make sense.

File: extensions/steve/steve.go
Comment (Josh-Diamond) [reply extensions/steve/steve.go:20]: i guess it is very similar, but in the tests, having require.NoError or assert.NoError makes it more readable IMO.  

What do you want me to do about this?  I like it, but if its not getting merged, i'll nuke it 🙂

Comment (igomez06) [reply extensions/steve/steve.go:20]: What's the difference of doing
```
require.NoError(client.Delete)
require.NoError(DeleteSteveResource)
```
?
It really does not add much. You other functions make sense, this does not. Remove this and you got my approval.

Comment (Josh-Diamond) [reply extensions/steve/steve.go:20]: Screenshot 2024-04-26 at 10 56 58 AM

Comment (Josh-Diamond) [reply extensions/steve/steve.go]: updated as requested 👍

**Review by igomez06 — APPROVED** [review]
LGTM

**Review by markusewalker — APPROVED** [review]

**Comment by igomez06** [comment]
@Josh-Diamond there will be backports correct?

**Comment by Josh-Diamond** [comment]
@igomez06 yes, I was awaiting the review process for `v2.9` first, and once approvals were given I plan on drafting up the backports 👍

**Review by slickwarren — APPROVED** [review]

**Pull request merged**
